js/block.js

```diff
@@ -143,55 +153,127 @@ StepProto.run = function(scope){
 };
 window.WBStep = document.registerElement('wb-step', {prototype: StepProto});
 
-var variableLocalsToUpdate = null;
-var oldVariableName = '';
-
-function handleVariableFocus(evt){
-    // Gather all the locals so we can update them
-    // TODO: Cancel if the new variable name is already in scope
-    var input = evt.target;
-    if (!input.parentElement.hasAttribute('isvariable')){
+function updateVariable(evt){
+    var setVariableBlock = evt.detail;
+    // ignore variables not in the workspace (in scripts)
+    if (! dom.matches(setVariableBlock, 'wb-workspace *')){
         return;
     }
-    var parentContext = dom.closest(input, 'wb-contains');
-    var variableName = input.parentElement.getAttribute('value');
-    variableLocalsToUpdate = dom.findAll(parentContext, 'wb-expression[script="control.getVariable"]')
-                                          .filter(function(expr){
-        var value = dom.find(expr, 'wb-value');
-        return (value && value.getAttribute('value') === variableName);
-    });
+    // ignore blocks that are not setVariable
+    if (setVariableBlock.getAttribute('script') !== 'control.setVariable'){
+        return;
+    }
+    var valueBlock = evt.target;
+    var type = valueBlock.getAttribute('type');
+    setTypeOfVariable(setVariableBlock, type);
+    updateLocalInstances(setVariableBlock, type);
 }
+Event.on(workspace, 'editor:wb-added', 'wb-expression', updateVariable);
+
+function createLocalAssociation(evt){
+    var setVariableBlock = evt.target;
+    if (!setVariableBlock.hasAttribute('id')){
+        console.log('createLocalAssociation');
+        var id = randomId();
+        setVariableBlock.setAttribute('id', id);
+        var local = dom.find(setVariableBlock, 'wb-local [script="control.getVariable"]');
+        local.setAttribute('for', id);
+    }
+}
+Event.on(workspace, 'editor:wb-added', '[script="control.setVariable"]', createLocalAssociation);
+
+
+var oldVariableName = '';
+
+// function handleVariableFocus(evt){
+//     // Gather all the locals so we can update them
+//     var input = evt.target;
+//     var parent = dom.closest(input, '[script="control.setVariable"]');
+//     var parentContext = dom.closest(input, 'wb-contains');
+//     variableLocalsToUpdate = getVariablesToUpdate(parentContext, parent.id);
+// }
 
 function handleVariableInput(evt){
     // Actually change the locals while we update
     var input = evt.target;
```

Comment:
`parent` should be the `wb-step` and it should keep internal references to the input (and maybe the context, whenever it is added to the workspace?). The step itself should manage its own internal state.